packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/sass-language.ts

```diff
@@ -97,31 +103,49 @@ async function compileString(
             url,
             options: FileImporterWithRequestContextOptions,
           ): Promise<URL | null> => {
+            let resolved = resolutionCache.get(url);
+            if (resolved !== undefined) {
+              return resolved;
+            }
+
+            // Default to not found
+            resolved = null;
+
             const result = await resolveUrl(url, options);
             if (result.path) {
-              return pathToFileURL(result.path);
+              resolved = pathToFileURL(result.path);
+            } else {
+              // Check for package deep imports
+              const parts = url.split('/');
+              const hasScope = parts.length >= 2 && parts[0].startsWith('@');
+              const [nameOrScope, nameOrFirstPath, ...pathPart] = parts;
+              const packageName = hasScope ? `${nameOrScope}/${nameOrFirstPath}` : nameOrScope;
+
+              // Caching package root locations is particularly beneficial for `@material/*` packages
+              // which extensively use deep imports.
+              let packageRoot = packageRootCache.get(packageName);
+              if (packageRoot === undefined) {
+                const packageResult = await resolveUrl(packageName + '/package.json', options);
+                if (packageResult.path) {
+                  packageRoot = dirname(packageResult.path);
+                  packageRootCache.set(packageName, packageRoot);
+                }
+              }
```

Comment:
Suggestion: `const packageRoot = resolvePackageRoot(packageName);` would help separate concerns and reduce the mutability which readers need to follow.